libnmstate/state.py

```diff
@@ -264,3 +269,41 @@ def dict_update(origin_data, to_merge_data):
         else:
             origin_data[key] = val
     return origin_data
+
+
+class LinuxBridgeState(object):
+    DEFAULT_STP_ENABLED = True
+    DEFAULT_PORT_STP_HAIRPIN_MODE = False
+    DEFAULT_PORT_STP_PATH_COST = 100
+    DEFAULT_PORT_STP_PRIORITY = 32
+
+    DEFAULT_PORT_STP_OPTIONS = {
+        LinuxBridge.PORT_STP_HAIRPIN_MODE: DEFAULT_PORT_STP_HAIRPIN_MODE,
+        LinuxBridge.PORT_STP_PATH_COST: DEFAULT_PORT_STP_PATH_COST,
+        LinuxBridge.PORT_STP_PRIORITY: DEFAULT_PORT_STP_PRIORITY,
+    }
+
+    STP_OPTIONS = [
+        LinuxBridge.STP_PRIORITY,
+        LinuxBridge.STP_FORWARD_DELAY,
+        LinuxBridge.STP_HELLO_TIME,
+        LinuxBridge.STP_MAX_AGE,
+    ]
+
+    @staticmethod
+    def canonicalize(iface_state):
```

Comment:
What is the general idea of these kind of methods? AFAIU a canonical representation is an agreed representation that all equivalent states can be transformed to (see https://en.wikipedia.org/wiki/Canonicalization). For example:

```
  - name: linux-br0
    type: linux-bridge
    state: up
    bridge:
      options:
      port:
        - name: eth1
        - name: eth3
        - name: eth2
``` 

and

```
  - name: linux-br0
    type: linux-bridge
    state: up
    bridge:
      options:
      port:
        - name: eth1
        - name: eth2
        - name: eth3
```

are AFAIU equivalent states. Therefore canonicalization should either transform one of these states into the other or both into the same, canonical representation.

Since the method is currently not doing this, IMHO it should be renamed or it would need to be adjusted.

Also IMHO it would be nicer if a `LinuxBridgeState` object would describe the actual state and canonicalization would happen implicitly, so

```
bridge_a = LinuxBridgeState(ports=["eth3", "eth2", "eth1"])  # I am not saying that the constructor should look like this, but just demonstrate that there are two objects that just differ in the order of ports
bridge_b = LinuxBridgeState(ports=["eth1", "eth2", "eth3"])
assert bridge_a == bridge_b
assert bridge_a.json == bridge_b.json

# I guess to verify partial states, it might be nice to also support `in`

assert bridge_a in bridge_b
```

Reply:
That's too big for this bug fix PR.
But I will consider the `LinuxBridgeState.__init__()` to do the canonicalization and `LinuxBridgeState.to_dict()` for `State` to update its `iface_state`.